src/main/java/org/truffleruby/core/binding/BindingNodes.java

```diff
@@ -273,6 +294,23 @@ public Object localVariableSetCached(DynamicObject binding, String name, Object
             return writeLocalVariableNode.executeWrite(frame, value);
         }
 
+        @Specialization(guards = {
+                "name == cachedName",
+                "!hiddenVariable(cachedName)",
+                "getFrameDescriptor(binding) == cachedFrameDescriptor",
+                "cachedFrameSlot == null"
+        }, limit = "getCacheLimit()")
+        public Object localVariableSetNewCached(DynamicObject binding, String name, Object value,
+                @Cached("name") String cachedName,
+                @Cached("getFrameDescriptor(binding)") FrameDescriptor cachedFrameDescriptor,
+                @Cached("findFrameSlotOrNull(name, getTopFrame(binding))") FrameSlotAndDepth cachedFrameSlot,
+                @Cached("newFrameDescriptor(getContext(), name)") FrameDescriptor newDescriptor,
+                @Cached("findFrameSlotOrNull(name, newDescriptor)") FrameSlotAndDepth newFrameSlot,
+                @Cached("createWriteNode(newFrameSlot)") WriteFrameSlotNode writeLocalVariableNode) {
+            final MaterializedFrame frame = newExtrasFrame(binding, newDescriptor);
+            return writeLocalVariableNode.executeWrite(frame, value);
```

Comment:
Nice! This looks quite similar to https://github.com/graalvm/truffleruby/blob/42f65ca47d9799683364d44eda9492095992982b/src/main/java/org/truffleruby/language/objects/WriteObjectFieldNode.java#L90